pkg/liquidity-mining/test/GaugeMaxRelativeWeight.test.ts

```diff
@@ -0,0 +1,213 @@
+    const balTokenAdmin = await deploy('MockBalancerTokenAdmin', { args: [vault.address, BAL.address] });
+    const balMinter = await deploy('BalancerMinter', { args: [balTokenAdmin.address, gaugeController.address] });
+
+    // We won't be using the code that requires the VotingEscrowDelegationProxy, so we just use any address.
+    liquidityGaugeImplementation = await deploy('LiquidityGaugeV5', {
+      args: [balMinter.address, ANY_ADDRESS, adaptor.address],
+    });
+    liquidityGaugeFactory = await deploy('LiquidityGaugeFactory', { args: [liquidityGaugeImplementation.address] });
+    stakelessGaugeFactory = await deploy('SingleRecipientGaugeFactory', { args: [balMinter.address] });
+  });
+
+  sharedBeforeEach('deploy gauges', async () => {
+    liquidityGauge = await deployedAt(
+      'LiquidityGaugeV5',
+      await deployGauge(liquidityGaugeFactory, token0.address, 'GaugeCreated')
+    );
+    stakelessGauge = await deployedAt(
+      'SingleRecipientGauge',
+      await deployGauge(stakelessGaugeFactory, token0.address, 'SingleRecipientGaugeCreated')
+    );
+  });
+
+  sharedBeforeEach('set up permissions', async () => {
+    const action = await actionId(adaptor, 'set_max_relative_weight', liquidityGaugeImplementation.interface);
+    await vault.grantPermissionsGlobally([action], admin);
+  });
+
+  async function deployGauge(gaugeFactory: Contract, poolAddress: string, eventName: string): Promise<string> {
+    const tx = await gaugeFactory.create(poolAddress);
+    const event = expectEvent.inReceipt(await tx.wait(), eventName);
+
+    return event.args.gauge;
+  }
+
+  async function setNewMaxRelativeWeight(maxRelativeWeight: BigNumber) {
+    const calldata = gauge.interface.encodeFunctionData('set_max_relative_weight', [maxRelativeWeight]);
+    return adaptor.connect(admin).performAction(gauge.address, calldata);
+  }
+
+  describe('LiquidityGaugeV5', () => {
+    beforeEach('use liquidity gauge', () => {
+      gauge = liquidityGauge;
+    });
+    itTestsMaxRelativeWeightForGauge();
+  });
+
+  describe('StakelessGauge', () => {
+    beforeEach('use stakeless gauge', () => {
+      gauge = stakelessGauge;
+    });
+    itTestsMaxRelativeWeightForGauge();
+  });
+
+  function itTestsMaxRelativeWeightForGauge() {
+    describe('set_maximum_relative_weight', () => {
+      const defaultMaxRelativeWeight = fp(1.0);
+      const newMaxRelativeWeight = fp(0.3);
+      const aboveAbsoluteMaxRelativeWeight = defaultMaxRelativeWeight.add(1);
+
+      context('without permissions', () => {
+        it('reverts', async () => {
+          await expect(gauge.set_max_relative_weight(newMaxRelativeWeight)).to.be.reverted;
+        });
+      });
+
+      context('with permissions', () => {
+        context('within the allowed range', () => {
+          it('sets max relative weight', async () => {
+            expect(await gauge.max_relative_weight()).to.be.eq(defaultMaxRelativeWeight);
```

Comment:
```suggestion
```

Reply:
actually I caught a bug with this line :grimacing: 

At first I had initialized the default in the gauge constructor, but since gauges are proxies that didn't work, and it made me realize I had to do it in the `initialize()` method.

Reply:
Ah good catch.

But that doesn't mean that the test belongs here though. You should have a set of tests for what the contract looks like after creation (e.g. that the constructor arguments were correctly applied). That'd let you spot another incorrect behavior, which is that the constructor is not emitting `MaxRelativeWeightChanged` events.

Reply:
New test added at the beginning (deployment).